File: backend/app/schemas/saved_property.py
"""
SavedProperty schemas for user's property portfolio management.
"""

from datetime import datetime
from decimal import Decimal
from enum import StrEnum
from typing import Any

from pydantic import BaseModel, Field

from app.schemas.deal_maker import DealMakerRecord


class PropertyStatus(StrEnum):
    """Status of a saved property."""

    WATCHING = "watching"
    ANALYZING = "analyzing"
    CONTACTED = "contacted"
    UNDER_CONTRACT = "under_contract"
    OWNED = "owned"
    PASSED = "passed"
    ARCHIVED = "archived"


# ===========================================
# Saved Property Schemas
# ===========================================


class SavedPropertyBase(BaseModel):
    """Base saved property fields."""

    nickname: str | None = Field(None, max_length=100)
    tags: list[str] | None = None
    color_label: str | None = Field(None, pattern="^(red|green|blue|yellow|purple|orange|gray)$")
    priority: int | None = Field(None, ge=1, le=5)
    notes: str | None = None


class SavedPropertyCreate(SavedPropertyBase):
    """Schema for saving a property."""

    # Property identification
    external_property_id: str | None = Field(None, max_length=100, description="ID from property service cache")
    zpid: str | None = Field(None, max_length=50, description="Zillow property ID")

    # Address (required)
    address_street: str = Field(..., max_length=255)
    address_city: str | None = Field(None, max_length=100)
    address_state: str | None = Field(None, max_length=10)
    address_zip: str | None = Field(None, max_length=20)
    full_address: str | None = Field(None, max_length=500)

    # Property data snapshot — capped to prevent oversized payloads.
    # A typical property snapshot is ~5-15 KB; 512 KB is generous headroom.
    property_data_snapshot: dict[str, Any] | None = Field(
        None,
        description="Full property data at time of save (max ~512 KB serialized)",
    )

    # Deal Maker Record - the central analysis data structure
    # If provided, will be stored as-is. If not, will be created from property data + defaults
    deal_maker_record: DealMakerRecord | None = Field(
        None,
        description="Central analysis record with property data + assumptions + adjustments",
    )

    # Initial status
    status: PropertyStatus = PropertyStatus.WATCHING

    @classmethod
    def _check_json_size(cls, v: dict | None, field_name: str, max_bytes: int = 524_288) -> dict | None:
        """Reject JSON payloads exceeding *max_bytes* (default 512 KB)."""
        if v is not None:
            import json

            size = len(json.dumps(v, default=str))
            if size > max_bytes:
                raise ValueError(f"{field_name} payload is {size:,} bytes — max allowed is {max_bytes:,} bytes")
        return v

    from pydantic import model_validator

    @model_validator(mode="after")
    def _enforce_json_limits(self) -> "SavedPropertyCreate":
        """Validate that JSON blob fields don't exceed size limits."""
        self._check_json_size(self.property_data_snapshot, "property_data_snapshot")
        if self.deal_maker_record is not None:
            import json

            size = len(json.dumps(self.deal_maker_record.model_dump(mode="json"), default=str))
            if size > 524_288:
                raise ValueError(f"deal_maker_record payload is {size:,} bytes — max allowed is 524,288 bytes")
        return self


class SavedPropertyUpdate(SavedPropertyBase):
    """Schema for updating a saved property."""

    status: PropertyStatus | None = None
    display_order: int | None = None

    # Custom value adjustments (DEPRECATED - use deal_maker_record)
    custom_purchase_price: Decimal | None = Field(None, ge=0, le=100_000_000)
    custom_rent_estimate: Decimal | None = Field(None, ge=0, le=1_000_000)
    custom_arv: Decimal | None = Field(None, ge=0, le=100_000_000)
    custom_rehab_budget: Decimal | None = Field(None, ge=0, le=100_000_000)
    custom_daily_rate: Decimal | None = Field(None, ge=0, le=100_000)
    custom_occupancy_rate: Decimal | None = Field(None, ge=0, le=1)

    # Custom assumptions per strategy (DEPRECATED - use deal_maker_record)
    custom_assumptions: dict[str, Any] | None = None

    # Deal Maker Record - the central analysis data structure
    deal_maker_record: DealMakerRecord | None = Field(
        None,
        description="Central analysis record - replaces custom_assumptions",
    )

    from pydantic import model_validator

    @model_validator(mode="after")
    def _enforce_json_limits(self) -> "SavedPropertyUpdate":
        """Validate that JSON blob fields don't exceed 512 KB."""
        import json

        _max = 524_288
        for field_name in ("custom_assumptions",):
            val = getattr(self, field_name, None)
            if val is not None:
                size = len(json.dumps(val, default=str))
                if size > _max:
                    raise ValueError(f"{field_name} payload is {size:,} bytes — max is {_max:,}")
        if self.deal_maker_record is not None:
            size = len(json.dumps(self.deal_maker_record.model_dump(mode="json"), default=str))
            if size > _max:
                raise ValueError(f"deal_maker_record payload is {size:,} bytes — max is {_max:,}")
        return self


class SavedPropertySummary(BaseModel):
    """Summary view of saved property for list views."""

    id: str
    address_street: str
    address_city: str | None
    address_state: str | None
    address_zip: str | None
    nickname: str | None
    status: PropertyStatus
    tags: list[str] | None
    color_label: str | None
    priority: int | None

    # Quick metrics
    best_strategy: str | None
    best_cash_flow: Decimal | None = None
    best_coc_return: Decimal | None = None

    # Timestamps
    saved_at: datetime
    last_viewed_at: datetime | None
    updated_at: datetime

    class Config:
        from_attributes = True

    @property
    def display_name(self) -> str:
        """Return nickname or address."""
        return self.nickname or self.address_street


class SavedPropertyResponse(SavedPropertySummary):
    """Full saved property response."""

    user_id: str
    external_property_id: str | None
    zpid: str | None
    full_address: str | None

    # Full property data
    property_data_snapshot: dict[str, Any] | None

    # Custom adjustments (DEPRECATED - use deal_maker_record)
    custom_purchase_price: Decimal | None = None
    custom_rent_estimate: Decimal | None = None
    custom_arv: Decimal | None = None
    custom_rehab_budget: Decimal | None = None
    custom_daily_rate: Decimal | None = None
    custom_occupancy_rate: Decimal | None = None
    custom_assumptions: dict[str, Any] | None

    # Deal Maker Record - the central analysis data structure
    deal_maker_record: DealMakerRecord | None = Field(
        None, description="Central analysis record with property data + assumptions + user adjustments"
    )

    # Notes
    notes: str | None

    # Analytics cache
    last_analytics_result: dict[str, Any] | None
    analytics_calculated_at: datetime | None

    # Refresh timestamp
    data_refreshed_at: datetime | None

    # Related counts
    document_count: int = 0
    adjustment_count: int = 0

    class Config:
        from_attributes = True


# ===========================================
# Property Adjustment Schemas
# ===========================================


class PropertyAdjustmentCreate(BaseModel):
    """Schema for creating a property adjustment record."""

    adjustment_type: str = Field(..., description="Type: purchase_price, rent, arv, assumptions, status, notes")
    field_name: str | None = Field(None, description="Specific field changed within type")
    previous_value: Any | None = None
    new_value: Any | None = None
    reason: str | None = Field(None, description="User's explanation for the change")


class PropertyAdjustmentResponse(PropertyAdjustmentCreate):
    """Schema for adjustment response."""

    id: str
    property_id: str
    created_at: datetime

    class Config:
        from_attributes = True


# ===========================================
# Bulk Operations
# ===========================================


class BulkStatusUpdate(BaseModel):
    """Schema for bulk status update."""

    property_ids: list[str] = Field(..., min_length=1)
    status: PropertyStatus


class BulkTagUpdate(BaseModel):
    """Schema for bulk tag update."""

    property_ids: list[str] = Field(..., min_length=1)
    add_tags: list[str] | None = None
    remove_tags: list[str] | None = None

File: mobile/types/savedProperty.ts
/**
 * Saved Property types - matching backend schemas/saved_property.py exactly
 */

import { PropertyStatus, ColorLabel, StrategyType } from './api';
import { DealMakerRecord } from './dealMaker';

// ===========================================
// Saved Property Summary (list view)
// ===========================================
export interface SavedPropertySummary {
  id: string;
  address_street: string;
  address_city: string | null;
  address_state: string | null;
  address_zip: string | null;
  nickname: string | null;
  status: PropertyStatus;
  tags: string[] | null;
  color_label: ColorLabel | null;
  priority: number | null; // 1-5

  // Quick metrics
  best_strategy: StrategyType | null;
  best_cash_flow: number | null;
  best_coc_return: number | null;

  // Timestamps
  saved_at: string; // ISO datetime
  last_viewed_at: string | null;
  updated_at: string;
}

// ===========================================
// Saved Property Response (full detail)
// ===========================================
export interface SavedPropertyResponse extends SavedPropertySummary {
  user_id: string;
  external_property_id: string | null;
  zpid: string | null;
  full_address: string | null;

  // Full property data
  property_data_snapshot: Record<string, unknown> | null;

  // Custom adjustments (DEPRECATED - use deal_maker_record)
  custom_purchase_price: number | null;
  custom_rent_estimate: number | null;
  custom_arv: number | null;
  custom_rehab_budget: number | null;
  custom_daily_rate: number | null;
  custom_occupancy_rate: number | null;
  custom_assumptions: Record<string, unknown> | null;

  // Worksheet assumptions (DEPRECATED - use deal_maker_record)
  worksheet_assumptions: Record<string, unknown> | null;

  // Deal Maker Record - the central analysis data structure
  deal_maker_record: DealMakerRecord | null;

  // Notes
  notes: string | null;

  // Analytics cache
  last_analytics_result: Record<string, unknown> | null;
  analytics_calculated_at: string | null;

  // Refresh timestamp
  data_refreshed_at: string | null;

  // Related counts
  document_count: number;
  adjustment_count: number;
}

// ===========================================
// Saved Property Create Request
// ===========================================
export interface SavedPropertyCreate {
  // Property identification
  external_property_id?: string | null;
  zpid?: string | null;

  // Address (required)
  address_street: string;
  address_city?: string | null;
  address_state?: string | null;
  address_zip?: string | null;
  full_address?: string | null;

  // Property data snapshot
  property_data_snapshot?: Record<string, unknown> | null;

  // Deal Maker Record
  deal_maker_record?: DealMakerRecord | null;

  // Initial status
  status?: PropertyStatus;

  // Optional metadata
  nickname?: string | null;
  tags?: string[] | null;
  color_label?: ColorLabel | null;
  priority?: number | null;
  notes?: string | null;
}

// ===========================================
// Saved Property Update Request
// ===========================================
export interface SavedPropertyUpdate {
  nickname?: string | null;
  tags?: string[] | null;
  color_label?: ColorLabel | null;
  priority?: number | null;
  notes?: string | null;
  status?: PropertyStatus;
  display_order?: number;

  // Custom value adjustments (DEPRECATED - use deal_maker_record)
  custom_purchase_price?: number | null;
  custom_rent_estimate?: number | null;
  custom_arv?: number | null;
  custom_rehab_budget?: number | null;
  custom_daily_rate?: number | null;
  custom_occupancy_rate?: number | null;
  custom_assumptions?: Record<string, unknown> | null;

  // Worksheet assumptions (DEPRECATED - use deal_maker_record)
  worksheet_assumptions?: Record<string, unknown> | null;

  // Deal Maker Record - the central analysis data structure
  deal_maker_record?: DealMakerRecord | null;
}

// ===========================================
// Property Adjustment
// ===========================================
export interface PropertyAdjustmentCreate {
  adjustment_type: string;
  field_name?: string | null;
  previous_value?: unknown;
  new_value?: unknown;
  reason?: string | null;
}

export interface PropertyAdjustmentResponse extends PropertyAdjustmentCreate {
  id: string;
  property_id: string;
  created_at: string;
}

// ===========================================
// Bulk Operations
// ===========================================
export interface BulkStatusUpdate {
  property_ids: string[];
  status: PropertyStatus;
}

export interface BulkTagUpdate {
  property_ids: string[];
  add_tags?: string[] | null;
  remove_tags?: string[] | null;
}

export interface BulkDeleteRequest {
  property_ids: string[];
}

// ===========================================
// Portfolio Stats
// ===========================================
export interface PortfolioStats {
  total_properties: number;
  by_status: Record<PropertyStatus, number>;
  total_value: number | null;
  total_equity: number | null;
  total_monthly_cash_flow: number | null;
  average_coc_return: number | null;
}

// ===========================================
// Helper Functions
// ===========================================
export function getDisplayAddress(property: SavedPropertySummary): string {
  return property.nickname || property.address_street;
}

export function getShortAddress(property: SavedPropertySummary): string {
  return property.address_street;
}

export function getCityStateZip(property: SavedPropertySummary): string {
  const parts: string[] = [];
  if (property.address_city) parts.push(property.address_city);
  if (property.address_state) parts.push(property.address_state);
  if (property.address_zip) parts.push(property.address_zip);
  return parts.join(', ');
}
